streamz/tests/test_kafka.py

```diff
@@ -195,18 +206,16 @@ def test_kafka_dask_batch(c, s, w1, w2):
     ARGS = {'bootstrap.servers': 'localhost:9092',
             'group.id': 'streamz-test%i' % j}
     with kafka_service() as kafka:
+        kafka, TOPIC = kafka
         stream = Stream.from_kafka_batched(TOPIC, ARGS, asynchronous=True,
                                            dask=True)
+        out = stream.gather().sink_to_list()
         stream.start()
+        yield gen.sleep(5)  # this frees the loop while dask workers report in
```

Comment:
Would an `await_for` statement make more sense here?  Do we know what keeps the dask workers so long?

Reply:
I do not know what's happening in this time. I tried `yield await_for(lambda: c.ncores())`, but get
```distributed/distributed/utils_test.py:945, AssertionError```, an absolute timeout in `gen_cluster` and warnings like
```
distributed.core - INFO - Event loop was unresponsive in Worker for 10.98s.
```
i.e., the gen.sleep of a single seems to release where querying the cluster is actually blocking.

Reply:
Note that `c.ncores` will be a coroutine in async mode, so you would need to `yield` the result.

However, I also don't understand the need here.  Any workers that will arrive have arrived at the start of the function.  `gen_cluster` waits until everyone is here.

Reply:
Yes, I yielded for ncores - didn't work :|
Sorry, I don't know more, I can't think how to investigate this further. The following also fails, with the same timeout in gen_cluster (*not* in the test itself)
```
        t = time.time()
        while True:
            yield gen.sleep(0.2)
            ncore = yield c.ncores()
            if ncore:
                break
            assert time.time() - t < 10, "Timeout"
```

Reply:
I'm still confused about what this sleep is doing.  I would be *very* surprised if the workers weren't already up and connected at this point.

```python
In [1]: from distributed.utils_test import gen_cluster

In [2]: @gen_cluster(client=True)
   ...: def test_foo(c, s, a, b):
   ...:     assert len(s.workers) == 2
   ...:

In [3]: test_foo()
distributed.scheduler - INFO - Clear task state
distributed.scheduler - INFO -   Scheduler at:     tcp://127.0.0.1:52172
distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:52173
distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:52173
distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:52172
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          1
distributed.worker - INFO -                Memory:                   17.18 GB
distributed.worker - INFO -       Local Directory: /Users/mrocklin/workspace/distributed/dask-worker-space/worker-c59n7r0g
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:52174
distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:52174
distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:52172
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          2
distributed.worker - INFO -                Memory:                   17.18 GB
distributed.worker - INFO -       Local Directory: /Users/mrocklin/workspace/distributed/dask-worker-space/worker-btnrzcte
distributed.worker - INFO - -------------------------------------------------
distributed.scheduler - INFO - Register tcp://127.0.0.1:52173
distributed.scheduler - INFO - Register tcp://127.0.0.1:52174
distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:52173
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:52174
distributed.core - INFO - Starting established connection
distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:52172
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:52172
distributed.worker - INFO - -------------------------------------------------
distributed.core - INFO - Starting established connection
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Receive client connection: Client-3da15092-4b26-11e9-9923-acde48001122
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Remove client Client-3da15092-4b26-11e9-9923-acde48001122
distributed.scheduler - INFO - Remove client Client-3da15092-4b26-11e9-9923-acde48001122
distributed.scheduler - INFO - Close client connection: Client-3da15092-4b26-11e9-9923-acde48001122
distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:52173
distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:52174
distributed.scheduler - INFO - Scheduler closing...
distributed.scheduler - INFO - Scheduler closing all comms
distributed.scheduler - INFO - Remove worker tcp://127.0.0.1:52173
distributed.core - INFO - Removing comms to tcp://127.0.0.1:52173
distributed.batched - INFO - Batched Comm Closed: in <closed TCP>: Stream is closed
distributed.scheduler - INFO - Remove worker tcp://127.0.0.1:52174
distributed.core - INFO - Removing comms to tcp://127.0.0.1:52174
distributed.scheduler - INFO - Lost all workers
distributed.worker - INFO - Connection to scheduler broken.  Reconnecting...
distributed.worker - INFO - Connection to scheduler broken.  Reconnecting...
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO - -------------------------------------------------
```

We may still need a sleep here, but it would be good to know why.

Reply:
Agreed, but I don't know how to investigate.